File: 200-number-of-islands/200.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution
{
public:
    int DIR = 4;
    vector<int> dirx = {1, -1, 0, 0};
    vector<int> diry = {0, 0, 1, -1};
    vector<vector<bool>> _visited;
    vector<vector<char>> _grid;
    int m, n;
    int numIslands(vector<vector<char>> &grid)
    {
        m = grid.size();
        n = grid.empty() ? 0 : grid[0].size();
        if (n == 0)
            return 0;
        _grid = grid;
        _visited.assign(m, vector<bool>(n, false));
        int res = 0;
        for (int i = 0; i < m; ++i)
        {
            for (int j = 0; j < n; ++j)
            {
                if (_grid[i][j] == '1' && !_visited[i][j])
                {
                    dfs(i, j);
                    ++res;
                }
            }
        }
        return res;
    }

    bool valid(int x, int y)
    {
        return x >= 0 && x < m && y >= 0 && y < n;
    }
    void dfs(int x, int y)
    {
        if (!valid(x, y) || _grid[x][y] == '0' || _visited[x][y])
            return;

        _visited[x][y] = true;
        for (int i = 0; i < DIR; ++i)
        {
            int nx = x + dirx[i];
            int ny = y + diry[i];
            dfs(nx, ny);
        }
    }
};

int main(int argc, char **args)
{
    Solution sol = Solution();
    vector<vector<char>> grid_1 = {{'1', '1', '1', '1', '0'},
                                   {'1', '1', '0', '1', '0'},
                                   {'1', '1', '0', '0', '0'},
                                   {'0', '0', '0', '0', '0'}};
    cout << sol.numIslands(grid_1) << endl;
    vector<vector<char>> grid_2 = {{'1', '1', '0', '0', '0'},
                                   {'1', '1', '0', '0', '0'},
                                   {'0', '0', '1', '0', '0'},
                                   {'0', '0', '0', '1', '1'}};
    cout << sol.numIslands(grid_2) << endl;
    return 0;
}

File: 200-number-of-islands/200.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

from typing import List


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        def valid(x: int, y: int):
            return x >= 0 and x < len(grid) and y >= 0 and y < len(grid[0])

        visited: List[List[bool]] = [[False for i in range(len(grid[0]))]
                                     for j in range(len(grid))]

        def dfs(x: int, y: int):
            if not valid(x, y) or visited[x][y] or grid[x][y] == '0':
                return

            visited[x][y] = True
            for (a, b) in [(1, 0), (0, 1), (-1, 0), (0, -1)]:
                nx, ny = x + a, y + b
                dfs(nx, ny)

        res = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1' and not visited[i][j]:
                    res += 1
                    dfs(i, j)

        return res


if __name__ == '__main__':
    sol = Solution()
    grid_val = [["1", "1", "1", "1", "0"], ["1", "1", "0", "1", "0"],
                ["1", "1", "0", "0", "0"], ["0", "0", "0", "0", "0"]]
    print(sol.numIslands(grid_val))
    grid_val = [["1", "1", "0", "0", "0"], ["1", "1", "0", "0", "0"],
                ["0", "0", "1", "0", "0"], ["0", "0", "0", "1", "1"]]
    print(sol.numIslands(grid_val))
